aten/src/ATen/core/autocast/autocast_mode.cpp

```diff
@@ -0,0 +1,65 @@
+#include <ATen/ATen.h>
+#include <ATen/core/autocast/autocast_mode.h>
+#include <c10/core/impl/LocalDispatchKeySet.h>
+
+#include <stdexcept>
+#include <memory>
+
+namespace at {
+namespace autocast {
+
+/// thread_local is a feature that is not enabled by Caffe2 mobile
+/// build (e.g. iOS). Therefore, we only provide `at::AutocastMode`
+/// when we are not in mobile build or when FEATURE_TORCH_MOBILE
+/// is on.
+#if !defined(C10_MOBILE) || defined(FEATURE_TORCH_MOBILE)
+bool AutocastMode::is_enabled() {
+  return c10::impl::tls_is_dispatch_key_included(DispatchKey::AutocastTensorId);
+}
+
+void AutocastMode::set_enabled(bool new_enabled) {
+  c10::impl::tls_set_dispatch_key_included(DispatchKey::AutocastTensorId, new_enabled);
+}
+
+void clear_cache();
+int increment_nesting();
+int decrement_nesting();
```

Comment:
There isn't really any reason to forward declare these functions, right?

Reply:
To put it differently, these functions should be made available in a header and then included here, rather than manually declaring them as has been done here.

Reply:
I forward declared them so I wouldn't need an additional header file.  Making a header file seemed overkill since autocast_mode.cpp is the only consumer of `clear_cache` and `increment/decrement_nesting`.

Reply:
I don't even need the `AutocastMode` wrapper struct.  I only structured it that way because I was cargo culting `grad_mode.h` and `grad_mode.cpp`.  As I think we discussed earlier, I don't understand why grad_mode.* use the GradMode wrapper struct, but assumed they had some good intention.  Maybe the author felt it was cleaner than declaring a separate namespace.  If so, that's irrelevant to me because I've got an autocast namespace.

I could dispense with AutocastMode entirely and move the bodies of the interface functions from autocast_mode.cpp into register_autocast_ops.cpp.  Should I do that?

Reply:
The main reason to put them in a header file is to make sure you keep them consistent if they mangle to the same symbol. The header can be included from the definition site. Admittedly this is low risk with C++ mangling, but I don't want to think hard about whether or not it is *impossible* to happen.

But yeah, if you just get rid of AutocastMode entirely that seems pretty good. I think the only countervailing question is whether or not you expect AMP to be usable from C++, but that's something that can always be solved at a later point in time.

Reply:
Are you saying forward declarations might cause different (and potentially divergent across compile units) mangling than declarations in a header?  I didn't realize that was a concern, I thought headers were basically copy pasted into compile units.  But sure I'll get rid of AutocastMode for now.

I definitely want to add C++ API support in the future, but if C++ support requires any changes they can be added later.  Right now everything is implementation details.